extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableAdapter.java

```diff
@@ -633,46 +655,312 @@ private static TableDefinition fromSchema(
+                        "partitioning columns.");
+            }
+        }
+    }
+
+    private static void verifyPartitionPathsIfAppending(
+            final org.apache.iceberg.Table icebergTable,
+            final Collection<String> partitionPaths) {
+        if (icebergTable.spec().isPartitioned() && partitionPaths.isEmpty()) {
+            throw new IllegalArgumentException("Cannot write data to a partitioned table without partition paths.");
+        } else if (!icebergTable.spec().isPartitioned() && !partitionPaths.isEmpty()) {
+            throw new IllegalArgumentException("Cannot write data to an un-partitioned table with partition paths.");
+        }
+    }
+
+    private static class CompletedParquetWrite {
+        private final URI destination;
+        private final long numRows;
+        private final long numBytes;
+
+        private CompletedParquetWrite(final URI destination, final long numRows, final long numBytes) {
+            this.destination = destination;
+            this.numRows = numRows;
+            this.numBytes = numBytes;
+        }
+    }
+
+    @NotNull
+    private List<CompletedParquetWrite> writeParquet(
+            @NotNull final PartitionSpec partitionSpec,
+            @NotNull final List<Table> dhTables,
+            @NotNull final List<String> partitionPaths,
+            @NotNull final IcebergParquetWriteInstructions writeInstructions) {
+        // Build the parquet instructions
+        final List<CompletedParquetWrite> parquetFilesWritten = new ArrayList<>(dhTables.size());
+        final ParquetInstructions.OnWriteCompleted onWriteCompleted =
+                (destination, numRows, numBytes) -> parquetFilesWritten
+                        .add(new CompletedParquetWrite(destination, numRows, numBytes));
+
+        // Remove all partitioning columns from the table definition before writing parquet files
+        TableDefinition parquetTableDefinition = writeInstructions.tableDefinition()
+                .orElseThrow(() -> new IllegalArgumentException("Table definition expected"));
+        if (partitionSpec.isPartitioned()) {
+            parquetTableDefinition = TableDefinition.of(
+                    parquetTableDefinition.getColumnStream()
+                            .filter(columnDefinition -> !columnDefinition.isPartitioning())
+                            .collect(Collectors.toList()));
+        }
+        final ParquetInstructions parquetInstructions = writeInstructions.toParquetInstructions(
+                onWriteCompleted, parquetTableDefinition, table.schema().idToName());
+
+        // Write the data to parquet files
+        int count = 0;
+        for (int idx = 0; idx < dhTables.size(); idx++) {
+            final Table dhTable = dhTables.get(idx);
+            if (dhTable.numColumns() == 0) {
+                // Skip writing empty tables with no columns
+                continue;
+            }
+            final String fileName = String.format(
```

Comment:
I wonder if we should be using the Iceberg TableLocation provider thing for this.